src/software/simulated_tests/simulated_test_fixture.py

```diff
@@ -164,13 +167,27 @@ def runner(
             time_elapsed_s += tick_duration_s
 
             if self.thunderscope:
-                time.sleep(tick_duration_s)
+                time.sleep(
+                    tick_duration_s - primitive_block_time
+                    if primitive_block_time >= 0
+                    else 0
+                )
```

Comment:
I think the `sleep` can be moved to the end of the loop, and the block_time/iteration_run_time should be measured from the start of the loop, to the right before the `sleep` call. Though let me know if you think that will run into weird edge cases.

Also, the logic for sleeping I think should be something like:
```python
                if tick_duration_s > iteration_run_time:
                    time.sleep(
                        tick_duration_s - iteration_run_time
                    )
```
to avoid sleeping for negative value (not sure if there will be an exception or not) and also, `primitive_block_time`/`iteration_run_time` should always be positive.

Reply:
Though if you made these changes, just double check them by running a couple of decently long tests to make sure that it looks like the simulation is running in realtime.

Reply:
yeah that makes sense, my previous implementation just measured the time that it blocked once, but not the case if it times out and loops for a bit. i've made the change now